onmt/modules/WeightNorm.py

```diff
@@ -0,0 +1,225 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn import Parameter
+from torch.autograd import Variable
+
+
+def get_var_maybe_avg(namespace, var_name, training, polyak_decay):
+    ''' utility for retrieving polyak averaged params '''
+
+    # Update average
+    v = getattr(namespace, var_name)
+    v_avg = getattr(namespace, var_name + '_avg')
+    # if v_avg.sum() == 0:
+    #     v_avg.copy_(v.data)
+    # else:
+    v_avg -= (1 - polyak_decay) * (v_avg - v.data)
+
+    if training:
+        return v
+    else:
+        return Variable(v_avg)
+
+
+def get_vars_maybe_avg(namespace, var_names, training, polyak_decay):
+    ''' utility for retrieving polyak averaged params '''
+    vars = []
+    for vn in var_names:
+        vars.append(get_var_maybe_avg(
+            namespace, vn, training, polyak_decay))
+    return vars
+
+
+class WN_Linear(nn.Linear):
```

Comment:
Can you have better comments here? Maybe a paper citation?

Dumb question, but does this need to be implemented differently for each layer?

Reply:
yes, it does. Weight Normalization is a re-weight method not like Batch Normalization is a re-output method, so it can't be implemented like a activation follow the layer.